File: Leetcode/06_Stack & Queue/029_0225_implement-stack-using-queues.py
class MyStack(object):
    def __init__(self):
        self.q = []

    def push(self, x):
        self.q.append(x)

    def pop(self):
        for i in range(len(self.q)):
            cur = self.q.pop(0)
            if i == len(self.q): # 这里是等于len因为pop之后len已经-1了
                return cur
            self.q.append(cur)

    def top(self):
        for i in range(len(self.q)):
            cur = self.q.pop(0)
            self.q.append(cur)
        return cur
            

    def empty(self):
        return not self.q

File: Leetcode/06_Stack & Queue/029_0225_implement-stack-using-queues.go
// https://leetcode.cn/problems/implement-stack-using-queues
// e

package main

type MyStack struct {
	q []int
}

func Constructor() MyStack {
	return MyStack{}
}

func (s *MyStack) Push(x int) {
	s.q = append(s.q, x)
}

func (s *MyStack) Pop() (cur int) {
	n := len(s.q)
	for i := range n {
		cur = s.q[0]
		s.q = s.q[1:]
		if i == n-1 {
			return
		}
		s.q = append(s.q, cur)
	}
	return // only to avoid syntax error
}

func (s *MyStack) Top() (cur int) {
	n := len(s.q)
	for range n {
		cur = s.q[0]
		s.q = s.q[1:]
		s.q = append(s.q, cur)
	}
	return
}

func (s *MyStack) Empty() bool {
	return len(s.q) == 0
}

/**
 * Your MyStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * param_2 := obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.Empty();
 */
